src/ApiCore/Transport/RestTransport.php

```diff
@@ -156,15 +157,52 @@ private function getCallOptions(array $options)
         return $callOptions;
     }
 
+    /**
+     * @param \Exception $ex
+     * @return ApiException
+     */
     private function convertToApiException(\Exception $ex)
     {
-        $res = (string) $ex->getResponse()->getBody();
-        $rObj = (object) json_decode($res, true)['error'];
-        // Overwrite the HTTP Status Code with the RPC code, derived from the "status" field.
-        $rObj->code = ApiStatus::rpcCodeFromStatus($rObj->status);
-        $rObj->metadata = property_exists($rObj, 'details') ? $rObj->details : null;
-        $rObj->details = $rObj->message;
-
-        return ApiException::createFromStdClass($rObj);
+        $res = $ex->getResponse();
+        $body = (string) $res->getBody();
+        if ($error = json_decode($body, true)['error']) {
+            // Overwrite the HTTP Status Code with the RPC code if it exists.
+            $basicMessage = $error['message'];
+            $status = $error['status'];
+            $metadata = isset($error['details']) ? $error['details'] : null;
+        } else {
+            // Only map HTTP status codes from Google\Rpc\Code which do not map
+            // to multiple gRPC codes (e.g. excluding 400, 409, and 500).
+            $httpToRpcCodes = [
+                401 => ApiStatus::UNAUTHENTICATED,
+                403 => ApiStatus::PERMISSION_DENIED,
+                404 => ApiStatus::NOT_FOUND,
+                429 => ApiStatus::RESOURCE_EXHAUSTED,
+                499 => ApiStatus::CANCELLED,
+                501 => ApiStatus::UNIMPLEMENTED,
+                503 => ApiStatus::UNAVAILABLE,
+                504 => ApiStatus::DEADLINE_EXCEEDED,
+            ];
+            $status = isset($httpToRpcCodes[$res->getStatusCode()])
+                ? $httpToRpcCodes[$res->getStatusCode()]
+                : ApiStatus::UNKNOWN;
+            $basicMessage = $body;
+            $metadata = null;
+        }
+
+        $code = ApiStatus::rpcCodeFromStatus($status);
+        $messageData = [
```

Comment:
The code from here until the end of the method duplicates code in `ApiException::createFromStdClass`. Can we share that code? Maybe a factory method on `ApiException` like:
```
public static function createFromApiResponse($basicMessage, $code, $metadata) { ... }
```
that could be used here and in the implementation of `createFromStdClass`?

Reply:
Sounds good, added this method to `ApiException`